### ISSUE TITLE
The number of features in data (208) is not the same as it was in training data (209)

### ISSUE THREAD

**Issue by SpicerSolutions** [PatWalters/sfi#12]
I've run into a problem trying out your SFI code.

Running the repo as is it get the error

The number of features in data (208) is not the same as it was in training data (200)

Yesterday I rebuilt the model (see attached). I ran that and recieved a result I expected but when I moved the model to a different PC I get 

The number of features in data (208) is not the same as it was in training data (209)

[32_model.txt](https://github.com/PatWalters/sfi/files/11449925/32_model.txt)

**Comment by SpicerSolutions** [comment]
After having played around a bit it appears that building the model with Python 3.11 and trying to run it on 3.10 doesn't work. Installed 3.11 and now working as expected, the exact reason I'm unsure of.

**Comment by PatWalters** [comment]
I don't think this is an issue with Python versions.  It probably has to do with the RDKit version and the default number of descriptors calculated by the RDKit.  It's important to ensure that the same version of the RDKit is used for training and inference on the LogD model.  I'll make a note of this in the README.

**Comment by Moocember** [comment]
Thanks for posting this Pat.

If anyone else encounters this issue, this work around works with any version of RDKit:
`    def predict(self, mol):
        model_feature_names ="MaxEStateIndex MinEStateIndex MaxAbsEStateIndex MinAbsEStateIndex qed MolWt HeavyAtomMolWt ExactMolWt NumValenceElectrons NumRadicalElectrons MaxPartialCharge MinPartialCharge MaxAbsPartialCharge MinAbsPartialCharge FpDensityMorgan1 FpDensityMorgan2 FpDensityMorgan3 BalabanJ BertzCT Chi0 Chi0n Chi0v Chi1 Chi1n Chi1v Chi2n Chi2v Chi3n Chi3v Chi4n Chi4v HallKierAlpha Ipc Kappa1 Kappa2 Kappa3 LabuteASA PEOE_VSA1 PEOE_VSA10 PEOE_VSA11 PEOE_VSA12 PEOE_VSA13 PEOE_VSA14 PEOE_VSA2 PEOE_VSA3 PEOE_VSA4 PEOE_VSA5 PEOE_VSA6 PEOE_VSA7 PEOE_VSA8 PEOE_VSA9 SMR_VSA1 SMR_VSA10 SMR_VSA2 SMR_VSA3 SMR_VSA4 SMR_VSA5 SMR_VSA6 SMR_VSA7 SMR_VSA8 SMR_VSA9 SlogP_VSA1 SlogP_VSA10 SlogP_VSA11 SlogP_VSA12 SlogP_VSA2 SlogP_VSA3 SlogP_VSA4 SlogP_VSA5 SlogP_VSA6 SlogP_VSA7 SlogP_VSA8 SlogP_VSA9 TPSA EState_VSA1 EState_VSA10 EState_VSA11 EState_VSA2 EState_VSA3 EState_VSA4 EState_VSA5 EState_VSA6 EState_VSA7 EState_VSA8 EState_VSA9 VSA_EState1 VSA_EState10 VSA_EState2 VSA_EState3 VSA_EState4 VSA_EState5 VSA_EState6 VSA_EState7 VSA_EState8 VSA_EState9 FractionCSP3 HeavyAtomCount NHOHCount NOCount NumAliphaticCarbocycles NumAliphaticHeterocycles NumAliphaticRings NumAromaticCarbocycles NumAromaticHeterocycles NumAromaticRings NumHAcceptors NumHDonors NumHeteroatoms NumRotatableBonds NumSaturatedCarbocycles NumSaturatedHeterocycles NumSaturatedRings RingCount MolLogP MolMR fr_Al_COO fr_Al_OH fr_Al_OH_noTert fr_ArN fr_Ar_COO fr_Ar_N fr_Ar_NH fr_Ar_OH fr_COO fr_COO2 fr_C_O fr_C_O_noCOO fr_C_S fr_HOCCN fr_Imine fr_NH0 fr_NH1 fr_NH2 fr_N_O fr_Ndealkylation1 fr_Ndealkylation2 fr_Nhpyrrole fr_SH fr_aldehyde fr_alkyl_carbamate fr_alkyl_halide fr_allylic_oxid fr_amide fr_amidine fr_aniline fr_aryl_methyl fr_azide fr_azo fr_barbitur fr_benzene fr_benzodiazepine fr_bicyclic fr_diazo fr_dihydropyridine fr_epoxide fr_ester fr_ether fr_furan fr_guanido fr_halogen fr_hdrzine fr_hdrzone fr_imidazole fr_imide fr_isocyan fr_isothiocyan fr_ketone fr_ketone_Topliss fr_lactam fr_lactone fr_methoxy fr_morpholine fr_nitrile fr_nitro fr_nitro_arom fr_nitro_arom_nonortho fr_nitroso fr_oxazole fr_oxime fr_para_hydroxylation fr_phenol fr_phenol_noOrthoHbond fr_phos_acid fr_phos_ester fr_piperdine fr_piperzine fr_priamide fr_prisulfonamd fr_pyridine fr_quatN fr_sulfide fr_sulfonamd fr_sulfone fr_term_acetylene fr_tetrazole fr_thiazole fr_thiocyan fr_thiophene fr_unbrch_alkane fr_urea"
        model_feature_names = model_feature_names.split(' ')
        
        all_feature_names = self.descriptors.desc_names
        
        # For each model feature, find the corresponding index in the full list of features
        feature_indices = []
        for feature_name in model_feature_names:
            feature_indices.append(all_feature_names.index(feature_name))
        
        # Extract the features from the molecule
        features = self.descriptors.calc_mol(mol).reshape(1, -1)
        
        # Select only the features we need
        features = features[:, feature_indices]

        return self.mdl.predict(features)[0]`

**Comment by Moocember** [comment]
Thanks for posting this Pat.

If anyone else encounters this issue, this adaptation of the predict function from [logd_predictor.py](https://github.com/PatWalters/sfi/blob/main/logd_predictor.py) works  with any version of RDKit:
```
def predict(self, mol,model_file_name='model_logd.txt'):
        # Set model_feature_names from the model file (line 8). This is simply a string containing the order of features.       
        with open(model_file_name) as f:
            for i, line in enumerate(f):
                if i == 7:
                    model_feature_names = line.strip().split('=')[1]
                    break
        
        model_feature_names = model_feature_names.split(' ')
        
        all_feature_names = self.descriptors.desc_names
        
        # For each model feature, find the corresponding index in the full list of features
        feature_indices = []
        for feature_name in model_feature_names:
            feature_indices.append(all_feature_names.index(feature_name))
        
        # Extract the all features from the molecule
        features = self.descriptors.calc_mol(mol).reshape(1, -1)
        
        # Select only the features we need
        features = features[:, feature_indices]

        return self.mdl.predict(features)[0]
```
